src/Compiler/Utilities/TypeHashing.fs

```diff
@@ -406,130 +406,183 @@ module StructuralUtilities =
                 | None ->
-                    if r.Rigidity = TyparRigidity.Rigid then
-                        TypeToken.Rigid typarId
-                    else
-                        TypeToken.Unsolved typarId
-
-            | TType_measure m -> yield! emitMeasure m
-        }
+                    if out.Count < 256 then
+                        if r.Rigidity = TyparRigidity.Rigid then
+                            out.Add(TypeToken.Rigid typarId)
+                        else
+                            out.Add(TypeToken.Unsolved typarId)
 
-    let private getTypeStructureOfStrippedType (ty: TType) =
+            | TType_measure m -> emitMeasure ctx m
 
-        let env =
-            {
-                typarMap = System.Collections.Generic.Dictionary<Stamp, int>()
-                emitNullness = false
-                stable = true
-            }
+    let private getTypeStructureOfStrippedTypeUncached (ty: TType) =
+        let ctx = getContext ()
+        emitTType ctx ty
 
-        let tokens = emitTType env ty |> Seq.truncate 256 |> Seq.toArray
+        let out = ctx.Tokens
 
         // If the sequence got too long, just drop it, we could be dealing with an infinite type.
-        if tokens.Length = 256 then PossiblyInfinite
-        elif not env.stable then Unstable tokens
-        else Stable tokens
+        if out.Count >= 256 then PossiblyInfinite
+        elif not ctx.Stable then Unstable(out.ToArray())
+        else Stable(out.ToArray())
 
     // Speed up repeated calls by memoizing results for types that yield a stable structure.
-    let private memoize =
+    let private getTypeStructureOfStrippedType =
         WeakMap.cacheConditionally
             (function
             | Stable _ -> true
             | _ -> false)
-            getTypeStructureOfStrippedType
+            getTypeStructureOfStrippedTypeUncached
+
+    // A cooldown counter for unsolved TTypes.
+    // If a TType is unsolved, don't try to compute its structure for the next few calls.
+    let private unstableTokensCounts =
+        System.Runtime.CompilerServices.ConditionalWeakTable<TType, System.Runtime.CompilerServices.StrongBox<int>>()
 
     let tryGetTypeStructureOfStrippedType ty =
-        match memoize ty with
-        | PossiblyInfinite -> ValueNone
-        | ts -> ValueSome ts
+        match unstableTokensCounts.TryGetValue(ty) with
+        | true, count when count.Value > 0 ->
+            count.Value <- count.Value - 1
+            ValueNone
+        | found, count ->
+            match getTypeStructureOfStrippedType ty with
+            | PossiblyInfinite ->
+                let count =
+                    if found then
+                        count
+                    else
+                        unstableTokensCounts.GetOrCreateValue(ty)
+
+                count.Value <- 10
+                ValueNone
+            | Unstable _ as ts ->
+                let count =
+                    if found then
```

Comment:
If after cooldown and another attempt we still don't get a solved type it may be better to blacklist the given TType.